pkg/deploy/registry/deploylog/rest.go

```diff
@@ -0,0 +1,259 @@
+func (r *REST) NewGetOptions() (runtime.Object, bool, string) {
+	return &deployapi.DeploymentLogOptions{}, false, ""
+}
+
+// New creates an empty DeploymentLog resource
+func (r *REST) New() runtime.Object {
+	return &deployapi.DeploymentLog{}
+}
+
+// Get returns a streamer resource with the contents of the deployment log
+func (r *REST) Get(ctx kapi.Context, name string, opts runtime.Object) (runtime.Object, error) {
+	deployLogOpts, ok := opts.(*deployapi.DeploymentLogOptions)
+	if !ok {
+		return nil, errors.NewBadRequest("did not get an expected options.")
+	}
+
+	// Fetch deploymentConfig
+	config, err := r.Deploy.GetDeploymentConfig(ctx, name)
+	if err != nil {
+		return nil, errors.NewNotFound("deploymentConfig", name)
+	}
+	// See latest version; if 0, there are no deployments for this config
+	desiredVersion := config.LatestVersion
+	if desiredVersion == 0 {
+		return nil, errors.NewBadRequest(fmt.Sprintf("no deployment exists for deploymentConfig %q", config.Name))
+	}
+
+	// Support retrieving logs for older deployments
+	switch {
+	case deployLogOpts.Version == -1:
+		// Latest
+	case deployLogOpts.Version <= 0 || deployLogOpts.Version > config.LatestVersion:
+		// Invalid version
+		return nil, errors.NewBadRequest(fmt.Sprintf("invalid version for deploymentConfig %q: %d", config.Name, deployLogOpts.Version))
+	default:
+		desiredVersion = deployLogOpts.Version
+	}
+
+	// List deployments
+	deployments, err := r.Deploy.ListDeployments(ctx, deployutil.ConfigSelector(config.Name))
+	if err != nil {
+		return nil, err
+	}
+	if len(deployments.Items) == 0 {
+		return nil, errors.NewInternalError(fmt.Errorf("cannot find deployments for deploymentConfig %q", config.Name))
+	}
+	sort.Sort(deployutil.DeploymentsByLatestVersionDesc(deployments.Items))
+
+	// Find desired deployment
+	var target *kapi.ReplicationController
+	for _, deployment := range deployments.Items {
+		if desiredVersion == deployutil.DeploymentVersionFor(&deployment) {
+			target = &deployment
+			break
+		}
+	}
+	if target == nil {
+		desiredDeployment := deployutil.VersionedLabelForDeploymentConfig(config, desiredVersion)
+		glog.V(3).Infof("Cannot find desired deployment (%s) for %s", desiredDeployment, config.Name)
+		return nil, errors.NewNotFound("deployment", desiredDeployment)
+	}
+
+	// Check for deployment status; if it is new or pending, we will wait for it. If it is complete,
+	// the deployment completed successfully and the deployer pod will be deleted so we will return a
+	// success message. If it is running or failed, retrieve the log from the deployer pod.
+	status := deployutil.DeploymentStatusFor(target)
+	switch status {
+	case deployapi.DeploymentStatusNew, deployapi.DeploymentStatusPending:
+		if deployLogOpts.NoWait {
+			glog.V(4).Infof("Deployment %s is in %s state. No logs to retrieve yet.", deployutil.LabelForDeployment(target), status)
+			return &genericrest.LocationStreamer{}, nil
```

Comment:
I agree it's kind of an ambiguous state. I see both sides and am not sure which is more correct. Pedantically if the pod isn't being scheduled then its logs necessarily can't be found. If `NotFound` is being reused for both a missing deployment *and* missing logs for a found deployment in a `running` state, then we're already in a spot where `NotFound` is contextual. Is there some other error which makes more sense for missing deployment/pending logs so that `NotFound` can be narrowly scoped to "we found a running deployment and there should be logs available but there aren't"? Or is there a better error for that case so that `NotFound` can be used for the other cases?

I don't feel really strongly.